src/uct/cuda/gdr_copy/gdr_copy_md.c

```diff
@@ -30,77 +61,377 @@ static ucs_status_t uct_gdr_copy_md_query(uct_md_h md, uct_md_attr_t *md_attr)
+        ucs_error("gdr_pin_buffer Failed. length :%lu ", length);
+        return UCS_ERR_IO_ERROR;
+    }
+    if (mh == 0) {
+        ucs_error("gdr_pin_buffer Failed. length :%lu ", length);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    if (gdr_map(md->gdrcpy_ctx, mh, &bar_ptr, length) !=0) {
+        ucs_error("gdr_map failed. length :%lu ", length);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    if (gdr_get_info(md->gdrcpy_ctx, mh, &info) != 0) {
+        ucs_error("gdr_get_info failed. ");
+        return UCS_ERR_IO_ERROR;
+    }
+    mem_hndl->mh        = mh;
+    mem_hndl->info      = info;
+    mem_hndl->bar_ptr   = bar_ptr;
+    mem_hndl->reg_size  = length;
+
+    ucs_debug("registered memory:%p..%p length:%lu info.va:0x%"PRIx64" bar_ptr:%p",
+              address, address + length, length, info.va, bar_ptr);
+
+    return UCS_OK;
+}
+
+static ucs_status_t uct_gdr_copy_mem_dereg_internal(uct_md_h uct_md, uct_gdr_copy_mem_t *mem_hndl)
+{
+
+    uct_gdr_copy_md_t *md = ucs_derived_of(uct_md, uct_gdr_copy_md_t);
+
+    if (gdr_unmap(md->gdrcpy_ctx, mem_hndl->mh, mem_hndl->bar_ptr, mem_hndl->reg_size) !=0) {
+        ucs_error("gdr_unmap Failed. unpin_size:%lu ", mem_hndl->reg_size);
+        return UCS_ERR_IO_ERROR;
+    }
+    if (gdr_unpin_buffer(md->gdrcpy_ctx, mem_hndl->mh) !=0) {
+        ucs_error("gdr_unpin_buffer failed ");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ucs_debug("deregistered memorory. info.va:0x%"PRIx64" bar_ptr:%p",
+              mem_hndl->info.va, mem_hndl->bar_ptr);
+    return UCS_OK;
+}
+
+static ucs_status_t uct_gdr_copy_mem_reg(uct_md_h uct_md, void *address, size_t length,
                                          unsigned flags, uct_mem_h *memh_p)
 {
-    ucs_status_t rc;
-    uct_mem_h * mem_hndl = NULL;
+    uct_gdr_copy_mem_t *mem_hndl = NULL;
+    size_t reg_size;
+    void *ptr;
+    ucs_status_t status;
 
-    mem_hndl = ucs_malloc(sizeof(void *), "gdr_copy handle for test passing");
+    mem_hndl = ucs_malloc(sizeof(uct_gdr_copy_mem_t), "gdr_copy handle");
     if (NULL == mem_hndl) {
         ucs_error("Failed to allocate memory for gni_mem_handle_t");
-        rc = UCS_ERR_NO_MEMORY;
-        goto mem_err;
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    reg_size = (length + GPU_PAGE_SIZE - 1) & GPU_PAGE_MASK;
+    ptr = (void *) ((uintptr_t)address & GPU_PAGE_MASK);
+
+    status = uct_gdr_copy_mem_reg_internal(uct_md, ptr, reg_size, 0, mem_hndl);
+    if (status != UCS_OK) {
+        free(mem_hndl);
+        return status;
     }
+
     *memh_p = mem_hndl;
     return UCS_OK;
- mem_err:
-    return rc;
 }
 
-static ucs_status_t uct_gdr_copy_mem_dereg(uct_md_h md, uct_mem_h memh)
+static ucs_status_t uct_gdr_copy_mem_dereg(uct_md_h uct_md, uct_mem_h memh)
 {
-    ucs_free(memh);
-    return UCS_OK;
+    uct_gdr_copy_mem_t *mem_hndl = memh;
+    ucs_status_t status;
+
+    status = uct_gdr_copy_mem_dereg_internal(uct_md, mem_hndl);
+    free(mem_hndl);
```

Comment:
ucs_free
also if status !=OK don't release the handle. or give a warning and release it.